### PR TITLE
SCUMM: support hebrew fan-translations (text are written right-to-left)

### PR DESCRIPTION
**Pull request by BLooperZ**
This adds support for writing texts right to left for hebrew fan translations
(by Hebrew Adventure Group).

Translations are complete for most games.
But I prefer to wait and add the detection entries after final versions are released
(after proofreading and playtesting).

Thanks.

Games already translated:
The Secret of Monkey Island
Indiana Jones and the Fate of Atlantis
Day of the Tentacle
Sam & Max Hit the Road
Full Throttle
The Dig
The Curse of Monkey Island

(Zak McKracken and the Alien Mindbenders is also  translated but is not affected by this changes)

### CODE REVIEW

**Comment by tsoliman** [comment]
Some commit messages don't conform and some commits are "junk" that shouldn't be kept (e.g. the git push test).
You'll want to rebase/cleanup those before a merge happens.
Thanks.

(Note: this isn't a code review and I'm not a SCUMM engine dev)

**Comment by BLooperZ** [comment]
@tsoliman Thank you, fixed.

File: engines/scumm/string.cpp [hunk 1]
```diff
@@ -175,7 +179,7 @@ void ScummEngine_v6::drawBlastTexts() {
 				// Some localizations may override colors
 				// See credits in Chinese COMI
 				if (_game.id == GID_CMI &&	_language == Common::ZH_TWN &&
-				      c == '^' && (buf == _blastTextQueue[i].text + 1)) {
+					  c == '^' && (buf == _blastTextQueue[i].text + 1)) {
```
Comment (sev-) [inline engines/scumm/string.cpp]: Unnecessary change.

File: engines/scumm/string.cpp [hunk 2]
```diff
@@ -454,6 +461,88 @@ bool ScummEngine::newLine() {
 	return true;
 }
 
+void ScummEngine::fakeBidiString(char *ltext, bool ignoreVerb) {
+	int ll = 0;
```
Comment (sev-) [inline engines/scumm/string.cpp]: It would be useful if you describe the purpose of this method, and perhaps even what is going on, in the comments.

File: engines/scumm/string.cpp [hunk 3]
```diff
@@ -454,6 +461,88 @@ bool ScummEngine::newLine() {
...
+	if (_game.id == GID_INDY4 && ltext[ll] == 127) {
+		ll++;
+	}
+	while (ltext[ll] == -1) {
+		ll += 4;
+	}
+	int ipos = 0;
+	int start = 0;
+	char* text = ltext + ll;
```
Comment (sev-) [inline engines/scumm/string.cpp]: should be `char *`, the same thing below.

File: engines/scumm/string.cpp [hunk 4]
```diff
@@ -454,6 +461,88 @@ bool ScummEngine::newLine() {
...
+	char* current = text;
+	while(1) {
```
Comment (sev-) [inline engines/scumm/string.cpp]: `while (1)`

File: engines/scumm/string.cpp [hunk 5]
```diff
@@ -454,6 +461,88 @@ bool ScummEngine::newLine() {
...
+			// ignore the line break for verbs texts
+			if (ignoreVerb && (*(current + 1) ==  8)) {
+				*(current + 1) = *current;
+				*current = 8;
+				ipos += 2;
+				current += 2;
+				continue;
+			}
+
+			char buff[384];
```
Comment (sev-) [inline engines/scumm/string.cpp]: These are quite big for the stack. I recommend to `malloc` these two buffers  at the method start and free on the exit.

File: engines/scumm/string.cpp [hunk 6]
```diff
@@ -542,7 +631,7 @@ void ScummEngine::CHARSET_1() {
 		return;
 
 	if ((_game.version <= 6 && _haveMsg == 1) ||
-	    (_game.version == 7 && _haveMsg != 1)) {
+		(_game.version == 7 && _haveMsg != 1)) {
```
Comment (sev-) [inline engines/scumm/string.cpp]: Unnecessary change.

File: engines/scumm/string.cpp [hunk 7]
```diff
@@ -929,6 +1031,30 @@ void ScummEngine::drawString(int a, const byte *msg) {
...
 		_charset->_left -= _charset->getStringWidth(a, buf) / 2;
+	} else if (_game.version >= 4 && _game.version < 7 && _game.id != GID_SAMNMAX && _language == Common::HE_ISR) {
+		if (_game.id != GID_INDY4 || buf[0] == 127) {
+			int ll = 0;
+			if (_game.id == GID_INDY4 && buf[0] == 127) {
+				buf[0] = 32;
+				ll++;
+			}
+
+			byte* ltext = buf;
```
Comment (sev-) [inline engines/scumm/string.cpp]: `byte *ltext`

File: engines/scumm/string.cpp [hunk 8]
```diff
@@ -969,6 +1095,24 @@ void ScummEngine::drawString(int a, const byte *msg) {
 			case 8:
 				if (_charset->_center) {
 					_charset->_left = _charset->_startLeft - _charset->getStringWidth(a, buf + i);
+				} else if (_game.version >= 4 && _game.version < 7 && _language == Common::HE_ISR) {
+					int ll = 0;
+					byte* ltext = buf + i;
```
Comment (sev-) [inline engines/scumm/string.cpp]: `byte *ltext`

File: engines/scumm/verbs.cpp [hunk 9]
```diff
@@ -1010,6 +1011,9 @@ void ScummEngine_v7::drawVerb(int verb, int mode) {
 		_charset->setCurID(vs->charset_nr);
 
 		// Compute the text rect
+		if (_language == Common::HE_ISR /* && !(_game.id == GID_FT && (_charset->getCurID() == 6 || _charset->getCurID() == 7))*/) {
```
Comment (sev-) [inline engines/scumm/verbs.cpp]: Is the commented out part going to be used in the final version?

File: engines/scumm/string.cpp [hunk 10]
```diff
@@ -454,6 +461,88 @@ bool ScummEngine::newLine() {
...
+		}
+		if (*current) {
+			ipos++;
+			current++;
+			continue;
+		}
+		break;
+	}
+	if (!ignoreVerb && _game.id == GID_INDY4 && ltext[0] == 127) {
+		ltext[start + ipos + ll] = (char) 128;
```
Comment (sev-) [inline engines/scumm/string.cpp]: `(char)128`, or I would rather write `'\x80'`

File: engines/scumm/verbs.cpp (hunk 9 above)
Comment (BLooperZ) [reply engines/scumm/verbs.cpp]: After checking again, I think it is safe to remove the commented out code (the final version will work without it)

**Comment by BLooperZ** [comment]
Probably not very soon, I'm afraid.

The fake BiDi mechanism used here is custom made for SCUMM needs. it assumes all text should be written right-to-left while preserving left-to-right ordering for several formats of numbers (needed for variables) + some escaping of control codes.
(other game engines are mostly ok with just reversing the text of each line).
(The use of simple processing actually helps the translation, as it makes the ordering predictable).

The launcher GUI probably needs real BiDi processing, as it can have RTL texts (translated texts) alongside LTR text (such as paths, shaders and version info).

**Comment by sev-** [comment]
Thanks, merging.

**Comment by BLooperZ** [comment]
It seems the fake BiDi function does not work on system in which char is unsigned by default.
I'll soon update the code to use byte instead.
